src/main/java/com/conveyal/osmlib/NodeTracker.java

```diff
@@ -1,84 +1,78 @@
 package com.conveyal.osmlib;
 
 import com.google.common.collect.Maps;
+import org.roaringbitmap.RoaringBitmap;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.HashMap;
 import java.util.Map;
 
 /** 
  * A sparse bit set capable of handling 64-bit int indexes (like OSM IDs).
- * Each block is 64 64-bit longs. At 512 bytes long, it tracks 4096 IDs.
  * Node numbers in OSM tend to be contiguous so maybe the blocks should be bigger.
- * Disabled debug statements have no measurable effect on speed (verified).
  *
  * MapDB TreeSets are much faster than MapDB HashSets, but in-memory NodeTrackers are
  * much faster than MapDB TreeSets.
+ *
+ * To save space, this uses RoaringBitmaps. RoaringBitmaps currently only support 32 bit keys
+ * so we use multiple RoaringBitmaps containing the low 32 bits, stored in a map from the high
+ * 32 bits to a roaringbitmap. Since the OSM IDs are concentrated towards the bottom of the long
+ * space (i.e. they only need, so far, one more than an int provides), only a few blocks are used.
  */
 public class NodeTracker {
-
     private static final Logger LOG = LoggerFactory.getLogger(NodeTracker.class);
-    
-    /** Lowest 6 bits, index into a single 64-bit long int. */
-    private static final long LOW_MASK = 0x3F; 
-    private static final int LOW_SHIFT = 6;
-    
-    /** Lowest 12 bits, bit position within block. */
-    private static final long MID_MASK = 0xFFF;
-    
-    /** Highest 52 bits, the key for the block. */
-    private static final long BLOCK_MASK = ~MID_MASK;
-    
-    /** Each block is 64 64-bit longs. */ 
-    private Map<Long, long[]> blocks = Maps.newHashMap();
 
-    private int index(long x) {
-        return (int)((x & MID_MASK) >> LOW_SHIFT);
-    }
+    private Map<Integer, RoaringBitmap> blocks = new HashMap<>();
 
-    private int bit(long x) {
-        return (int)(x & LOW_MASK);
+    public void add(long x) {
+        int high = highIndex(x);
+        int low = lowIndex(x);
+
+        if (!blocks.containsKey(high)) {
```

Comment:
Typically rather than checking whether a map contains a key and then inserting, I would do the get() and then create and insert an empty element if the get() returns null. This way you only do one hash table look up in the case where the key is already present.